File: php-backend/src/Services/MappingUtils.php
<?php

declare(strict_types=1);

namespace App\Services;

/**
 * 报表行（field_mapping）辅助：code / 展示名 / 分组。
 * mapping 为关联数组（含 logical_field_code / logical_field_name / parts）。
 */
final class MappingUtils
{
    public static function mappingLineCode(array $m): string
    {
        if (!empty($m['line_code'])) {
            return (string) $m['line_code'];
        }
        if (!empty($m['logical_field_code'])) {
            return (string) $m['logical_field_code'];
        }
        return 'line_' . $m['id'];
    }

    public static function mappingLabel(array $m): string
    {
        if (!empty($m['label'])) {
            return (string) $m['label'];
        }
        if (!empty($m['logical_field_name'])) {
            return (string) $m['logical_field_name'];
        }
        return '指标' . $m['id'];
    }

    public static function isManualLine(array $m): bool
    {
        return strtolower((string) ($m['line_type'] ?? '')) === 'manual';
    }

    public static function isPerOrderLine(array $m): bool
    {
        return strtolower((string) ($m['line_type'] ?? '')) === 'per_order';
    }

    public static function isRatioLine(array $m): bool
    {
        return strtolower((string) ($m['line_type'] ?? '')) === 'ratio';
    }

    public static function isRefComputeLine(array $m): bool
    {
        $parts = $m['parts'] ?? [];
        if (!$parts) {
            return false;
        }
        foreach ($parts as $p) {
            if (trim((string) ($p['ref_field_code'] ?? '')) === '') {
                return false;
            }
        }
        return true;
    }

    public static function isFormulaLine(array $m): bool
    {
        if (self::isManualLine($m)) {
            return false;
        }
        $lineType = strtolower((string) ($m['line_type'] ?? ''));
        if ($lineType === 'formula') {
            return true;
        }
        if ($lineType === 'fetch') {
            return false;
        }
        $hasExpr = trim((string) ($m['expression'] ?? '')) !== '';
        $hasParts = !empty($m['parts']);
        $hasLegacy = !empty($m['sheet_name']) && !empty($m['column_header']);
        return $hasExpr && !$hasParts && !$hasLegacy;
    }

    /** 日报字段展示分类：placeholder | review | sample | compute | fetch | per_order | ratio | formula */
    public static function fieldDisplayType(array $m, ?string $lineCode = null): string
    {
        $code = $lineCode ?? self::mappingLineCode($m);
        $label = self::mappingLabel($m);
        if (self::isPerOrderLine($m)) {
            return 'per_order';
        }
        if (self::isRatioLine($m)) {
            return 'ratio';
        }
        if (self::isManualLine($m)) {
            return 'placeholder';
        }
        if (!empty($m['parts'])) {
            return self::isRefComputeLine($m) ? 'compute' : 'fetch';
        }
        if (in_array($label, MeichongRules::MANUAL_FILL_LABELS, true)) {
            return 'placeholder';
        }
        if (in_array($code, MeichongRules::REVIEW_IMPORT_CODES, true)) {
            return 'review';
        }
        if (in_array($code, MeichongRules::SAMPLE_IMPORT_CODES, true)) {
            return 'sample';
        }
        if (in_array($code, MeichongRules::PENDING_FILE_CODES, true)) {
            return 'placeholder';
        }
        if (self::isFormulaLine($m)) {
            return 'formula';
        }
        return 'fetch';
    }

    public static function isFetchLine(array $m): bool
    {
        return self::fieldDisplayType($m) === 'fetch';
    }

    /**
     * 字段 code → 展示名（逻辑字段名 + 报表行 label）。
     * @param array[] $mappings
     * @param array[] $logicalFields
     * @return array<string, string>
     */
    public static function buildFieldLabelsMap(array $mappings, array $logicalFields): array
    {
        $out = [];
        foreach ($logicalFields as $lf) {
            $code = (string) ($lf['code'] ?? '');
            $name = (string) ($lf['name'] ?? '');
            if ($code !== '' && $name !== '') {
                $out[$code] = $name;
            }
        }
        foreach ($mappings as $m) {
            $code = self::mappingLineCode($m);
            $label = self::mappingLabel($m);
            if ($code !== '' && $label !== '') {
                $out[$code] = $label;
            }
        }
        return $out;
    }

    /** @return string[] */
    public static function mappingSourceFileKeywords(array $m): array
    {
        $keywords = [];
        $parts = $m['parts'] ?? [];
        usort($parts, fn($a, $b) => ((int) ($a['sort_order'] ?? 0)) <=> ((int) ($b['sort_order'] ?? 0)));
        foreach ($parts as $p) {
            if (trim((string) ($p['ref_field_code'] ?? '')) !== '') {
                continue;
            }
            $kw = trim((string) ($p['source_file_keyword'] ?? ''));
            if ($kw !== '') {
                $keywords[$kw] = true;
            }
            foreach ($p['sources'] ?? [] as $s) {
                if (!is_array($s)) {
                    continue;
                }
                $sk = trim((string) ($s['source_file_keyword'] ?? ''));
                if ($sk !== '') {
                    $keywords[$sk] = true;
                }
            }
        }
        $keys = array_keys($keywords);
        sort($keys);
        return $keys;
    }

    private const AGG_LABELS = [
        'sum' => '求和',
        'count' => '计数',
        'count_distinct' => '去重计数',
        'sum_dedup' => '去重求和',
        'max_dedup' => '去重取最大',
        'avg' => '平均值',
    ];

    private static function resolveFileLabel(?string $keyword, ?array $fileLabels): string
    {
        $kw = trim((string) $keyword);
        if ($kw === '') {
            return '';
        }
        $labels = $fileLabels ?? [];
        return $labels[$kw] ?? $labels[strtolower($kw)] ?? $kw;
    }

    private static function sourceLoc(array $source, ?array $fileLabels): string
    {
        $fl = self::resolveFileLabel($source['source_file_keyword'] ?? null, $fileLabels);
        $col = trim((string) ($source['column_header'] ?? ''));
        if ($fl !== '' && $col !== '') {
            return "{$fl}.{$col}";
        }
        return $col !== '' ? $col : ($fl !== '' ? $fl : '未指定');
    }

    /** 列表页规则摘要：订单.Order Amount · 去重求和 */
    public static function partRuleBrief(array $part, ?array $fileLabels = null, ?array $fieldLabels = null): string
    {
        $ref = trim((string) ($part['ref_field_code'] ?? ''));
        if ($ref !== '') {
            $labels = $fieldLabels ?? [];
            return $labels[$ref] ?? $ref;
        }

        $sources = $part['sources'] ?? [];
        if ($sources) {
            $pieces = [];
            foreach ($sources as $i => $src) {
                $loc = self::sourceLoc(is_array($src) ? $src : [], $fileLabels);
                $pieces[] = $i ? "+ {$loc}" : $loc;
            }
            $srcText = implode(' ', $pieces);
        } else {
            $srcText = self::sourceLoc([
                'source_file_keyword' => $part['source_file_keyword'] ?? null,
                'column_header' => $part['column_header'] ?? null,
            ], $fileLabels);
        }

        $agg = trim((string) ($part['aggregation'] ?? 'sum'));
        $aggLabel = self::AGG_LABELS[$agg] ?? $agg;
        return "{$srcText} · {$aggLabel}";
    }

    /** 纳入日报结构的行（非基础取数字段）。 */
    public static function isReportLine(array $m): bool
    {
        return (int) ($m['sort_order'] ?? 0) > 0 || !empty($m['report_group']);
    }

    /** @param array[] $mappings */
    public static function reportDisplayMappings(array $mappings): array
    {
        $lines = array_values(array_filter($mappings, fn($m) => self::isReportLine($m)));
        usort($lines, fn($a, $b) => [(int) ($a['sort_order'] ?? 0), (int) $a['id']] <=> [(int) ($b['sort_order'] ?? 0), (int) $b['id']]);
        return $lines;
    }

    public static function defaultExpression(array $m): string
    {
        $expr = trim((string) ($m['expression'] ?? ''));
        if ($expr !== '') {
            return $expr;
        }
        $code = self::mappingLineCode($m);
        return '{field:' . $code . '}';
    }

    /** @param array<string, bool>|array<int, string> $used 引用传递的已用 code 集合（code => true） */
    public static function slugLineCode(string $text, array &$used): string
    {
        $lower = mb_strtolower(trim($text));
        $ascii = preg_replace('/[^a-z0-9_]+/', '_', $lower) ?? '';
        $ascii = trim($ascii, '_');
        if ($ascii === '' || in_array($ascii, ['line', 'r_line'], true)) {
            $digest = substr(md5($text !== '' ? $text : 'line'), 0, 8);
            $base = "rpt_$digest";
        } else {
            $base = substr($ascii, 0, 40);
            if (ctype_digit($base[0])) {
                $base = "r_$base";
            }
        }
        $name = $base;
        $i = 2;
        while (isset($used[$name])) {
            $name = "{$base}_{$i}";
            $i++;
        }
        $used[$name] = true;
        return $name;
    }

    private const ROW_FILTER_HINTS = [
        'nonempty' => '非空',
        'empty' => '为空',
        'eq' => '=',
        'ne' => '≠',
        'in' => '∈',
        'not_in' => '∉',
        'contains' => '含',
        'not_contains' => '不含',
        'starts_with' => '开头',
        'ends_with' => '结尾',
        'gt' => '>',
        'gte' => '≥',
        'lt' => '<',
        'lte' => '≤',
        'between' => '介于',
    ];

    /** @return string[] */
    public static function partRuleHints(array $part): array
    {
        $hints = [];
        $dateCol = trim((string) ($part['date_filter_column'] ?? ''));
        if ($dateCol !== '') {
            $hints[] = "{$dateCol}=日报";
        }
        foreach ($part['row_filters'] ?? [] as $cond) {
            $col = trim((string) ($cond['column'] ?? ''));
            $op = $cond['op'] ?? 'eq';
            $values = $cond['values'] ?? [];
            if (is_string($values) || is_int($values) || is_float($values)) {
                $values = [$values];
            }
            $label = self::ROW_FILTER_HINTS[$op] ?? $op;
            if (in_array($op, ['nonempty', 'empty'], true)) {
                if ($col !== '') {
                    $hints[] = "{$col}{$label}";
                }
            } elseif (in_array($op, ['eq', 'ne', 'contains', 'not_contains', 'starts_with', 'ends_with', 'gt', 'gte', 'lt', 'lte'], true)) {
                $val = isset($values[0]) ? trim((string) $values[0]) : '';
                if ($col !== '' && $val !== '') {
                    $hints[] = "{$col}{$label}{$val}";
                }
            } elseif (in_array($op, ['in', 'not_in', 'between'], true) && $col !== '') {
                $vals = implode('、', array_map('strval', array_slice($values, 0, 3)));
                if ($vals !== '') {
                    $hints[] = "{$col}{$label}{$vals}";
                }
            }
        }
        $adv = [];
        if (!empty($part['exclude_sample'])) {
            $adv[] = '排除样品';
        }
        if (!empty($part['exclude_review'])) {
            $adv[] = '排除刷单';
        }
        if (!empty($part['join_to_orders'])) {
            $adv[] = '关联主表';
        }
        if (!empty($part['only_sample'])) {
            $adv[] = '仅样品';
        }
        if ($adv) {
            $hints[] = implode('、', $adv);
        }
        return $hints;
    }
}

File: app/services/mapping_utils.py
"""报表行（field_mapping）辅助：code / 展示名 / 分组。"""

from __future__ import annotations

import hashlib
import re

from app.models import FieldMapping

_LINE_CODE_RE = re.compile(r"[^a-z0-9_]+")


def mapping_line_code(mapping: FieldMapping) -> str:
    if mapping.line_code:
        return mapping.line_code
    if mapping.logical_field and mapping.logical_field.code:
        return mapping.logical_field.code
    return f"line_{mapping.id}"


def mapping_label(mapping: FieldMapping) -> str:
    if mapping.label:
        return mapping.label
    if mapping.logical_field and mapping.logical_field.name:
        return mapping.logical_field.name
    return f"指标{mapping.id}"


def mapping_source_file_keywords(mapping: FieldMapping) -> list[str]:
    """取数规则块直接绑定的来源文件 keyword（不含字段复用链）。"""
    keywords: set[str] = set()
    for p in sorted(mapping.parts or [], key=lambda x: x.sort_order):
        if p.ref_field_code:
            continue
        kw = (p.source_file_keyword or "").strip()
        if kw:
            keywords.add(kw)
        for s in p.sources or []:
            if not isinstance(s, dict):
                continue
            sk = (s.get("source_file_keyword") or "").strip()
            if sk:
                keywords.add(sk)
    return sorted(keywords)


def is_manual_line(mapping: FieldMapping) -> bool:
    return (mapping.line_type or "").lower() == "manual"


def is_per_order_line(mapping: FieldMapping) -> bool:
    """每单金额行：出报时 = per_order_amount × 单数（口径见 per_order_basis）。"""
    return (mapping.line_type or "").lower() == "per_order"


def is_ratio_line(mapping: FieldMapping) -> bool:
    """按比例行：出报时 = 复用字段(ratio_base_code) × ratio_percent%。"""
    return (mapping.line_type or "").lower() == "ratio"


# 每单金额单数口径
PER_ORDER_BASIS_VALID = "valid_orders"   # 当日去重有效订单数（下单口径）
PER_ORDER_BASIS_REVIEW = "review_orders"  # 刷单单数（不重复 Order ID）


def per_order_basis(mapping: FieldMapping) -> str:
    basis = (getattr(mapping, "per_order_basis", None) or "").strip()
    return PER_ORDER_BASIS_REVIEW if basis == PER_ORDER_BASIS_REVIEW else PER_ORDER_BASIS_VALID


def is_ref_compute_line(mapping: FieldMapping) -> bool:
    """日报行仅通过复用其他字段加减组合（parts 全是 ref_field_code）。"""
    parts = list(mapping.parts or [])
    if not parts:
        return False
    return all((getattr(p, "ref_field_code", None) or "").strip() for p in parts)


PLACEHOLDER_RULE_SUMMARY = "占位，导出后手工填写或上传文件"


def is_formula_line(mapping: FieldMapping) -> bool:
    if is_manual_line(mapping):
        return False
    if (mapping.line_type or "").lower() == "formula":
        return True
    if (mapping.line_type or "").lower() == "fetch":
        return False
    return bool((mapping.expression or "").strip()) and not mapping.parts and not (
        mapping.sheet_name and mapping.column_header
    )


def is_fetch_line(mapping: FieldMapping) -> bool:
    return field_display_type(mapping) == "fetch"


def field_display_type(mapping: FieldMapping, line_code: str | None = None) -> str:
    """日报字段展示分类：placeholder | review | sample | compute | fetch | per_order | ratio | formula。

    显式配置（每单金额/按比例/占位/取数或复用 parts）优先于 review/sample 等按 code 的启发判定，
    这样刷单物流费用等字段也能改用统一取数方式。
    """
    code = line_code or mapping_line_code(mapping)
    from app.services.meichong_rules import (
        MANUAL_FILL_LABELS,
        PENDING_FILE_CODES,
        REVIEW_IMPORT_CODES,
        SAMPLE_IMPORT_CODES,
    )
    label = mapping_label(mapping)
    # —— 显式配置优先 ——
    if is_per_order_line(mapping):
        return "per_order"
    if is_ratio_line(mapping):
        return "ratio"
    if is_manual_line(mapping):
        return "placeholder"
    if mapping.parts:
        return "compute" if is_ref_compute_line(mapping) else "fetch"
    # —— 无显式配置时，按 code / label 启发 ——
    if label in MANUAL_FILL_LABELS:
        return "placeholder"
    if code in REVIEW_IMPORT_CODES:
        return "review"
    if code in SAMPLE_IMPORT_CODES:
        return "sample"
    if code in PENDING_FILE_CODES:
        return "placeholder"
    if is_formula_line(mapping):
        return "formula"
    return "fetch"


def is_report_line(mapping: FieldMapping) -> bool:
    """纳入日报结构的行（非基础取数字段）。"""
    return (mapping.sort_order or 0) > 0 or bool(mapping.report_group)


def report_display_mappings(mappings: list[FieldMapping]) -> list[FieldMapping]:
    lines = [m for m in mappings if is_report_line(m)]
    return sorted(lines, key=lambda m: (m.sort_order or 0, m.id))


def default_expression(mapping: FieldMapping) -> str:
    if (mapping.expression or "").strip():
        return mapping.expression.strip()
    code = mapping_line_code(mapping)
    return f"{{field:{code}}}"


def slug_line_code(text: str, used: set[str]) -> str:
    ascii_part = _LINE_CODE_RE.sub("_", (text or "").lower().strip()).strip("_")
    if not ascii_part or ascii_part in ("line", "r_line"):
        digest = hashlib.md5((text or "line").encode("utf-8")).hexdigest()[:8]
        base = f"rpt_{digest}"
    else:
        base = ascii_part[:40]
        if base[0].isdigit():
            base = f"r_{base}"
    name = base
    i = 2
    while name in used:
        name = f"{base}_{i}"
        i += 1
    used.add(name)
    return name


_ROW_FILTER_HINTS: dict[str, str] = {
    "nonempty": "非空",
    "empty": "为空",
    "eq": "=",
    "ne": "≠",
    "in": "属于",
    "not_in": "不属于",
    "contains": "包含",
    "not_contains": "不含",
    "starts_with": "开头",
    "ends_with": "结尾",
    "gt": ">",
    "gte": "≥",
    "lt": "<",
    "lte": "≤",
    "between": "介于",
}


_AGG_LABELS: dict[str, str] = {
    "sum": "求和",
    "count": "计数",
    "count_distinct": "去重计数",
    "sum_dedup": "去重求和",
    "max_dedup": "去重取最大",
    "avg": "平均值",
}


def _agg_label(agg: str | None) -> str:
    key = (agg or "sum").strip()
    return _AGG_LABELS.get(key, key)


def _resolve_file_label(keyword: str | None, file_labels: dict[str, str] | None) -> str:
    kw = (keyword or "").strip()
    if not kw:
        return ""
    labels = file_labels or {}
    return labels.get(kw) or labels.get(kw.lower()) or kw


def _source_loc(source: dict, file_labels: dict[str, str] | None) -> str:
    fl = _resolve_file_label(source.get("source_file_keyword"), file_labels)
    col = (source.get("column_header") or "").strip()
    if fl and col:
        return f"{fl}.{col}"
    return col or fl or "未指定"


def _resolve_field_label(code: str | None, field_labels: dict[str, str] | None) -> str:
    c = (code or "").strip()
    if not c:
        return ""
    labels = field_labels or {}
    return labels.get(c) or c


def build_field_labels_map(mappings, logical_fields) -> dict[str, str]:
    """字段 code → 展示名（逻辑字段名 + 报表行 label）。"""
    out: dict[str, str] = {}
    for lf in logical_fields or []:
        code = getattr(lf, "code", None) or ""
        name = getattr(lf, "name", None) or ""
        if code and name:
            out[code] = name
    for m in mappings or []:
        code = mapping_line_code(m)
        label = mapping_label(m)
        if code and label:
            out[code] = label
    return out


def part_rule_brief(
    part,
    file_labels: dict[str, str] | None = None,
    field_labels: dict[str, str] | None = None,
) -> str:
    """列表页规则摘要：订单.Order Amount · 去重求和（详情见配置弹窗）。"""
    ref = (getattr(part, "ref_field_code", None) or "").strip()
    if ref:
        return _resolve_field_label(ref, field_labels)

    sources = getattr(part, "sources", None) or []
    if sources:
        pieces: list[str] = []
        for i, src in enumerate(sources):
            loc = _source_loc(src, file_labels)
            pieces.append(f"+ {loc}" if i else loc)
        src_text = " ".join(pieces)
    else:
        src_text = _source_loc(
            {
                "source_file_keyword": getattr(part, "source_file_keyword", None),
                "column_header": getattr(part, "column_header", None),
            },
            file_labels,
        )

    return f"{src_text} · {_agg_label(getattr(part, 'aggregation', None))}"


def _iter_part_row_filters(part) -> list[dict]:
    """行筛选：优先各来源列上的 row_filters，兼容旧 part 级配置。"""
    sources = getattr(part, "sources", None) or []
    if sources:
        out: list[dict] = []
        for src in sources:
            if isinstance(src, dict):
                out.extend(src.get("row_filters") or [])
        if out:
            return out
    return list(getattr(part, "row_filters", None) or [])


def part_rule_hints(part) -> list[str]:
    """取数 part 规则摘要后缀：日期列、行筛选等。"""
    hints: list[str] = []
    date_col = (getattr(part, "date_filter_column", None) or "").strip()
    if date_col:
        hints.append(f"{date_col}=日报")
    for cond in _iter_part_row_filters(part):
        col = str(cond.get("column", "")).strip()
        op = cond.get("op", "eq")
        values = cond.get("values", [])
        if isinstance(values, (str, int, float)):
            values = [values]
        label = _ROW_FILTER_HINTS.get(op, op)
        if op in ("nonempty", "empty"):
            if col:
                hints.append(f"{col}{label}")
        elif op in ("eq", "ne", "contains", "not_contains", "starts_with", "ends_with", "gt", "gte", "lt", "lte"):
            val = str(values[0]).strip() if values else ""
            if col and val:
                hints.append(f"{col}{label}{val}")
        elif op in ("in", "not_in", "between") and col:
            vals = "、".join(str(v) for v in values[:3])
            if vals:
                hints.append(f"{col}{label}{vals}")
    adv: list[str] = []
    if getattr(part, "exclude_sample", False):
        adv.append("排除样品")
    if getattr(part, "exclude_review", False):
        adv.append("排除刷单")
    if getattr(part, "exclude_same_day_refund", False):
        adv.append("排除当日退单")
    if getattr(part, "join_to_orders", False):
        adv.append("关联主表")
    if getattr(part, "only_sample", False):
        adv.append("仅样品")
    if adv:
        hints.append("、".join(adv))
    return hints
